File: mekaneck/virtualbrain/engine/poincare_computer.py
"""
PoincareComputer: Main Poincare computing engine.

Implements the Poincare computing paradigm for Virtual Brain simulation.
Programs specify completion conditions, and the engine navigates
backward through S-entropy space to satisfy constraints.
"""

from dataclasses import dataclass, field
from typing import Dict, List, Optional, Any
import numpy as np

from ..core.types import SCoord, MentalState, PartitionCoord
from ..core.constants import DEFAULT_COUPLING_STRENGTH, DEFAULT_N_OSCILLATORS
from ..operators.poincare_ops import COMPLETE, TARGET, EQUILIBRIUM
from ..operators.dynamics_ops import KURAMOTO, PHASE_LOCK, CASCADE, KuramotoState
from ..operators.neural_ops import evolve_mental_state, CONSCIOUSNESS


@dataclass
class ComputeResult:
    """Result of Poincare computation."""

    success: bool
    final_state: MentalState
    iterations: int
    convergence_history: List[float]
    trajectory: List[SCoord] = field(default_factory=list)
    metadata: Dict[str, Any] = field(default_factory=dict)


class PoincareComputer:
    """
    Main Poincare computing engine for Virtual Brain.

    Implements constraint satisfaction computing through
    equilibrium finding in categorical state space.

    The key insight is backward determination: specifying the
    completion state determines what the trajectory MUST have been.
    """

    def __init__(
        self,
        n_oscillators: int = DEFAULT_N_OSCILLATORS,
        coupling_strength: float = DEFAULT_COUPLING_STRENGTH,
        n_partition_levels: int = 10,
    ):
        """
        Initialize Poincare Computer.

        Args:
            n_oscillators: Number of Kuramoto oscillators
            coupling_strength: Default coupling strength K
            n_partition_levels: Max partition level n
        """
        self.n_oscillators = n_oscillators
        self.coupling_strength = coupling_strength
        self.n_partition_levels = n_partition_levels

        # Initialize Kuramoto network
        self.kuramoto_state = self._init_kuramoto()

        # State history for memory
        self.state_history: List[MentalState] = []

    def _init_kuramoto(self) -> KuramotoState:
        """Initialize Kuramoto oscillator network."""
        return KuramotoState.random(
            n_oscillators=self.n_oscillators,
            mean_frequency=10.0,
            frequency_std=2.0,
            coupling=self.coupling_strength,
        )

    def compute_consciousness(
        self,
        initial_state: MentalState,
        target_consciousness: float = 0.8,
        max_iterations: int = 1000,
        dt: float = 0.01,
    ) -> ComputeResult:
        """
        Compute state achieving target consciousness level.

        Uses Poincare completion to find state satisfying
        C = gamma * gamma_f = target.

        Args:
            initial_state: Starting mental state
            target_consciousness: Target consciousness level
            max_iterations: Maximum iterations
            dt: Time step

        Returns:
            ComputeResult with final state
        """
        convergence_history = []
        trajectory = []
        current = initial_state

        if current.s_coord:
            trajectory.append(current.s_coord)

        for iteration in range(max_iterations):
            # Evolve Kuramoto network
            self.kuramoto_state = KURAMOTO(self.kuramoto_state, dt)

            # Compute order parameter
            R, Psi = PHASE_LOCK(self.kuramoto_state.phases)

            # Update mental state with new coherence
            current = current.transition_to(new_gamma=R)

            # Evolve decay dynamics
            current = evolve_mental_state(current, dt)

            # Track convergence
            C = current.consciousness
            convergence_history.append(C)

            if current.s_coord:
                trajectory.append(current.s_coord)

            # Check convergence
            if abs(C - target_consciousness) < 0.01:
                return ComputeResult(
                    success=True,
                    final_state=current,
                    iterations=iteration,
                    convergence_history=convergence_history,
                    trajectory=trajectory,
                    metadata={"target": target_consciousness, "achieved": C},
                )

            # Adaptive coupling adjustment to reach target
            if C < target_consciousness:
                self.kuramoto_state = KuramotoState(
                    phases=self.kuramoto_state.phases,
                    natural_frequencies=self.kuramoto_state.natural_frequencies,
                    coupling_strength=self.kuramoto_state.coupling_strength * 1.01,
                )
            else:
                self.kuramoto_state = KuramotoState(
                    phases=self.kuramoto_state.phases,
                    natural_frequencies=self.kuramoto_state.natural_frequencies,
                    coupling_strength=self.kuramoto_state.coupling_strength * 0.99,
                )

        return ComputeResult(
            success=False,
            final_state=current,
            iterations=max_iterations,
            convergence_history=convergence_history,
            trajectory=trajectory,
            metadata={"target": target_consciousness, "achieved": current.consciousness},
        )

    def navigate_categorical_space(
        self,
        start: SCoord,
        target: SCoord,
        step_size: float = 0.1,
        max_steps: int = 100,
    ) -> List[SCoord]:
        """
        Navigate from start to target in S-entropy space.

        Implements O(log n) navigation via TARGET operator.

        Args:
            start: Starting S-entropy coordinate
            target: Target S-entropy coordinate
            step_size: Step size per iteration
            max_steps: Maximum navigation steps

        Returns:
            Path of SCoord from start to target
        """
        path = [start]
        current = start

        for _ in range(max_steps):
            if current.distance(target) < step_size:
                path.append(target)
                break

            current = TARGET(current, target, step_size)
            path.append(current)

        return path

    def apply_drug_perturbation(
        self,
        state: MentalState,
        drug_concentration: float,
        K_agg: float = 1e4,
        equilibration_steps: int = 100,
        dt: float = 0.01,
    ) -> MentalState:
        """
        Apply drug perturbation to mental state.

        Modifies coupling strength and evolves system to new equilibrium.

        Args:
            state: Current mental state
            drug_concentration: Drug concentration in M
            K_agg: Aggregation constant M^-1
            equilibration_steps: Steps to equilibrate
            dt: Time step

        Returns:
            Perturbed mental state after equilibration
        """
        # Modify coupling strength based on drug
        K_modified = self.coupling_strength * (1 + drug_concentration * K_agg)

        # Create modified Kuramoto state
        modified_kuramoto = KuramotoState(
            phases=self.kuramoto_state.phases,
            natural_frequencies=self.kuramoto_state.natural_frequencies,
            coupling_strength=K_modified,
        )

        # Evolve to new equilibrium
        for _ in range(equilibration_steps):
            modified_kuramoto = KURAMOTO(modified_kuramoto, dt)

        # Compute new coherence
        R, _ = PHASE_LOCK(modified_kuramoto.phases)

        # Update internal state
        self.kuramoto_state = modified_kuramoto

        return state.transition_to(new_gamma=R)

    def run_simulation(
        self,
        initial_state: MentalState,
        duration: float,
        dt: float = 0.01,
    ) -> List[MentalState]:
        """
        Run full simulation for given duration.

        Args:
            initial_state: Starting state
            duration: Simulation duration
            dt: Time step

        Returns:
            List of mental states over time
        """
        states = [initial_state]
        current = initial_state
        t = 0.0

        while t < duration:
            # Evolve Kuramoto
            self.kuramoto_state = KURAMOTO(self.kuramoto_state, dt)

            # Compute order parameter
            R, _ = PHASE_LOCK(self.kuramoto_state.phases)

            # Update state
            current = current.transition_to(new_gamma=R)
            current = evolve_mental_state(current, dt)

            states.append(current)
            self.state_history.append(current)
            t += dt

        return states

    def find_equilibrium(
        self,
        initial_state: MentalState,
        max_time: float = 100.0,
        dt: float = 0.01,
        tolerance: float = 1e-4,
    ) -> ComputeResult:
        """
        Find equilibrium state from initial conditions.

        Args:
            initial_state: Starting state
            max_time: Maximum simulation time
            dt: Time step
            tolerance: Convergence tolerance

        Returns:
            ComputeResult with equilibrium state
        """

        def dynamics(state, dt):
            self.kuramoto_state = KURAMOTO(self.kuramoto_state, dt)
            R, _ = PHASE_LOCK(self.kuramoto_state.phases)
            new_state = state.transition_to(new_gamma=R)
            return evolve_mental_state(new_state, dt)

        final_state, converged, history = EQUILIBRIUM(
            initial_state, dynamics, dt, max_time, tolerance
        )

        return ComputeResult(
            success=converged,
            final_state=final_state,
            iterations=len(history),
            convergence_history=history,
        )

    def dream_simulation(
        self,
        initial_state: MentalState,
        duration: float,
        dt: float = 0.01,
    ) -> List[MentalState]:
        """
        Simulate dream state (P_decay = 0).

        Thought trajectories are unbounded, navigating full S-space.

        Args:
            initial_state: Starting state
            duration: Dream duration
            dt: Time step

        Returns:
            List of mental states during dream
        """
        # Enter dream state
        dream_state = initial_state.enter_dream()

        return self.run_simulation(dream_state, duration, dt)

    def wake_transition(
        self,
        dream_state: MentalState,
        wake_duration: float = 1.0,
        dt: float = 0.01,
    ) -> List[MentalState]:
        """
        Simulate transition from dream to wake.

        Args:
            dream_state: State at end of dream
            wake_duration: Duration of wake transition
            dt: Time step

        Returns:
            List of mental states during transition
        """
        # Wake up
        wake_state = dream_state.wake(perception_level=0.1)

        states = [wake_state]
        current = wake_state
        t = 0.0

        while t < wake_duration:
            # Gradually increase perception
            new_p_decay = min(1.0, current.p_decay + dt)
            current = current.transition_to(new_p_decay=new_p_decay)

            # Evolve dynamics
            self.kuramoto_state = KURAMOTO(self.kuramoto_state, dt)
            R, _ = PHASE_LOCK(self.kuramoto_state.phases)
            current = current.transition_to(new_gamma=R)

            states.append(current)
            t += dt

        return states

    def reset(self):
        """Reset computer to initial state."""
        self.kuramoto_state = self._init_kuramoto()
        self.state_history = []

File: virtual_brain/crates/vb-engine/src/poincare_computer.rs
//! Poincare Computer: Main simulation engine.
//!
//! Implements the Poincare computing paradigm for consciousness simulation.

use serde::{Deserialize, Serialize};
use vb_core::constants::{DEFAULT_COUPLING_STRENGTH, DEFAULT_N_OSCILLATORS};
use vb_core::types::{MentalState, SCoord};
use vb_operators::{
    coherence, kuramoto, kuramoto_with_drug, navigate, KuramotoState,
};

/// Result of a computation.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ComputeResult {
    /// Whether computation succeeded
    pub success: bool,
    /// Final mental state
    pub final_state: MentalState,
    /// Number of iterations
    pub iterations: usize,
    /// Convergence history (consciousness over time)
    pub convergence_history: Vec<f64>,
    /// S-coordinate trajectory
    pub trajectory: Vec<SCoord>,
    /// Additional metadata
    pub metadata: std::collections::HashMap<String, String>,
}

/// Main Poincare computing engine.
#[derive(Debug, Clone)]
pub struct PoincareComputer {
    /// Kuramoto oscillator state
    kuramoto_state: KuramotoState,
    /// History of mental states
    state_history: Vec<MentalState>,
    /// Number of partition levels
    n_partition_levels: i32,
    /// Mean oscillator frequency
    mean_frequency: f64,
    /// Frequency standard deviation
    frequency_std: f64,
}

impl PoincareComputer {
    /// Create new Poincare computer.
    pub fn new(
        n_oscillators: usize,
        coupling_strength: f64,
        n_partition_levels: i32,
    ) -> Self {
        let mean_frequency = 10.0; // Hz
        let frequency_std = 1.0;

        Self {
            kuramoto_state: KuramotoState::random(
                n_oscillators,
                mean_frequency,
                frequency_std,
                coupling_strength,
            ),
            state_history: Vec::new(),
            n_partition_levels,
            mean_frequency,
            frequency_std,
        }
    }

    /// Get current Kuramoto coherence.
    pub fn current_coherence(&self) -> f64 {
        coherence(&self.kuramoto_state.phases)
    }

    /// Compute consciousness to target level.
    pub fn compute_consciousness(
        &mut self,
        initial_state: &MentalState,
        target_consciousness: f64,
        max_iterations: usize,
        dt: f64,
    ) -> ComputeResult {
        let mut state = initial_state.clone();
        let mut trajectory = Vec::new();
        let mut convergence = Vec::new();

        if let Some(s) = state.s_coord {
            trajectory.push(s);
        }

        for iter in 0..max_iterations {
            // Evolve Kuramoto
            self.kuramoto_state = kuramoto(&self.kuramoto_state, dt);
            let r = self.current_coherence();

            // Update mental state
            state = state.with_gamma(r);
            state = state.decay_perception(0.05, dt);
            state = state.decay_thought(0.1, dt);

            let c = state.consciousness();
            convergence.push(c);

            // Check convergence
            if (c - target_consciousness).abs() < 0.01 {
                return ComputeResult {
                    success: true,
                    final_state: state,
                    iterations: iter + 1,
                    convergence_history: convergence,
                    trajectory,
                    metadata: std::collections::HashMap::new(),
                };
            }

            // Adjust coupling to steer toward target
            if c < target_consciousness {
                self.kuramoto_state.coupling_strength *= 1.01;
            } else {
                self.kuramoto_state.coupling_strength *= 0.99;
            }
        }

        ComputeResult {
            success: false,
            final_state: state,
            iterations: max_iterations,
            convergence_history: convergence,
            trajectory,
            metadata: {
                let mut m = std::collections::HashMap::new();
                m.insert("reason".to_string(), "max_iterations".to_string());
                m
            },
        }
    }

    /// Navigate categorical space from start to target.
    pub fn navigate_categorical_space(
        &mut self,
        start: &SCoord,
        target: &SCoord,
        step_size: f64,
        max_steps: usize,
    ) -> Vec<SCoord> {
        navigate(start, target, step_size, max_steps)
    }

    /// Apply drug perturbation to dynamics.
    pub fn apply_drug_perturbation(
        &mut self,
        state: &MentalState,
        concentration: f64,
        k_agg: f64,
        dt: f64,
    ) -> MentalState {
        self.kuramoto_state = kuramoto_with_drug(&self.kuramoto_state, concentration, k_agg, dt);
        let r = self.current_coherence();
        state.with_gamma(r)
    }

    /// Run simulation for given duration.
    pub fn run_simulation(
        &mut self,
        initial_state: &MentalState,
        duration: f64,
        dt: f64,
    ) -> Vec<MentalState> {
        let n_steps = (duration / dt) as usize;
        let mut states = Vec::with_capacity(n_steps);
        let mut state = initial_state.clone();

        for _ in 0..n_steps {
            self.kuramoto_state = kuramoto(&self.kuramoto_state, dt);
            let r = self.current_coherence();

            state = state.with_gamma(r);
            state = state.decay_perception(0.05, dt);
            state = state.decay_thought(0.1, dt);
            state.timestamp += dt;

            states.push(state.clone());
        }

        self.state_history = states.clone();
        states
    }

    /// Find equilibrium state.
    pub fn find_equilibrium(
        &mut self,
        initial_state: &MentalState,
        max_time: f64,
        dt: f64,
        tolerance: f64,
    ) -> (MentalState, bool) {
        let mut state = initial_state.clone();
        let mut time = 0.0;
        let mut prev_c = state.consciousness();

        while time < max_time {
            self.kuramoto_state = kuramoto(&self.kuramoto_state, dt);
            let r = self.current_coherence();

            state = state.with_gamma(r);
            state = state.decay_perception(0.05, dt);
            state = state.decay_thought(0.1, dt);

            let c = state.consciousness();

            if (c - prev_c).abs() < tolerance {
                return (state, true);
            }

            prev_c = c;
            time += dt;
        }

        (state, false)
    }

    /// Simulate dream state.
    pub fn dream_simulation(
        &mut self,
        initial_state: &MentalState,
        duration: f64,
        dt: f64,
    ) -> Vec<MentalState> {
        let dream_state = initial_state.enter_dream();
        self.run_simulation(&dream_state, duration, dt)
    }

    /// Get state history.
    pub fn state_history(&self) -> &[MentalState] {
        &self.state_history
    }

    /// Reset Kuramoto state.
    pub fn reset_kuramoto(&mut self) {
        self.kuramoto_state = KuramotoState::random(
            self.kuramoto_state.n_oscillators(),
            self.mean_frequency,
            self.frequency_std,
            self.kuramoto_state.coupling_strength,
        );
    }
}

impl Default for PoincareComputer {
    fn default() -> Self {
        Self::new(DEFAULT_N_OSCILLATORS, DEFAULT_COUPLING_STRENGTH, 5)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_new() {
        let computer = PoincareComputer::new(50, 0.5, 5);
        assert!(computer.current_coherence() >= 0.0);
        assert!(computer.current_coherence() <= 1.0);
    }

    #[test]
    fn test_run_simulation() {
        let mut computer = PoincareComputer::new(50, 2.0, 5);
        let initial = MentalState::default();

        let states = computer.run_simulation(&initial, 1.0, 0.01);
        assert_eq!(states.len(), 100);
    }

    #[test]
    fn test_navigate() {
        let mut computer = PoincareComputer::default();
        let start = SCoord::origin();
        let target = SCoord::new(0.8, 0.8, 0.8).unwrap();

        let path = computer.navigate_categorical_space(&start, &target, 0.1, 100);
        assert!(path.len() > 1);
    }

    #[test]
    fn test_dream_simulation() {
        let mut computer = PoincareComputer::default();
        let initial = MentalState::default();

        let states = computer.dream_simulation(&initial, 0.5, 0.01);
        assert!(!states.is_empty());

        // In dream state, p_decay should be 0
        for state in &states {
            assert!(state.p_decay < 0.01);
        }
    }
}
